<a href="https://colab.research.google.com/github/AbuzarMohd/prime_course/blob/ML/ML_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [85]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.impute import  SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, precision_score, confusion_matrix, classification_report, accuracy_score, precision_score, recall_score, f1_score, mean_squared_error, r2_score
from sklearn.linear_model import LassoCV, RidgeCV


pd.set_option('display.max_column', None)

# Assignment 1

House Prine prediction using linaer regression

In [ ]:
df = pd.read_csv("/content/drive/MyDrive/AI Engineer Course by Apna College/datasets/HousePricePrediction.csv")
df.head()

,Id,MSSubClass,MSZoning,LotArea,LotConfig,BldgType,OverallCond,YearBuilt,YearRemodAdd,Exterior1st,BsmtFinSF2,TotalBsmtSF,SalePrice
0,0,60,RL,8450,Inside,1Fam,5,2003,2003,VinylSd,0.0,856.0,208500.0
1,1,20,RL,9600,FR2,1Fam,8,1976,1976,MetalSd,0.0,1262.0,181500.0
2,2,60,RL,11250,Inside,1Fam,5,2001,2002,VinylSd,0.0,920.0,223500.0
3,3,70,RL,9550,Corner,1Fam,5,1915,1970,Wd Sdng,0.0,756.0,140000.0
4,4,60,RL,14260,FR2,1Fam,5,2000,2000,VinylSd,0.0,1145.0,250000.0


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2919 entries, 0 to 2918
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Id            2919 non-null   int64  
 1   MSSubClass    2919 non-null   int64  
 2   MSZoning      2915 non-null   object 
 3   LotArea       2919 non-null   int64  
 4   LotConfig     2919 non-null   object 
 5   BldgType      2919 non-null   object 
 6   OverallCond   2919 non-null   int64  
 7   YearBuilt     2919 non-null   int64  
 8   YearRemodAdd  2919 non-null   int64  
 9   Exterior1st   2918 non-null   object 
 10  BsmtFinSF2    2918 non-null   float64
 11  TotalBsmtSF   2918 non-null   float64
 12  SalePrice     1460 non-null   float64
dtypes: float64(3), int64(6), object(4)
memory usage: 296.6+ KB


In [ ]:
df.isnull().sum()

,0
Id,0
MSSubClass,0
MSZoning,4
LotArea,0
LotConfig,0
BldgType,0
OverallCond,0
YearBuilt,0
YearRemodAdd,0
Exterior1st,1


In [ ]:
nums_imputer = SimpleImputer(strategy="mean")
nums_cols = ['BsmtFinSF2', 'TotalBsmtSF', 'SalePrice']
df[nums_cols] = nums_imputer.fit_transform(df[nums_cols])

cat_imputer = SimpleImputer(strategy="most_frequent")
cat_cols = ['MSZoning', 'Exterior1st']
df[cat_cols] = cat_imputer.fit_transform(df[cat_cols])

In [ ]:
df.isnull().sum()

,0
Id,0
MSSubClass,0
MSZoning,0
LotArea,0
LotConfig,0
BldgType,0
OverallCond,0
YearBuilt,0
YearRemodAdd,0
Exterior1st,0


In [ ]:
X = df.drop('SalePrice', axis=1)
y = df['SalePrice']

X = pd.get_dummies(X, columns=["MSZoning", "Exterior1st", "LotConfig", "BldgType"], drop_first=True)


In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=43)

In [ ]:
X['BldgType'].unique()
#

array(['1Fam', '2fmCon', 'Duplex', 'TwnhsE', 'Twnhs'], dtype=object)

In [ ]:
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [ ]:
lr_model = LinearRegression()

lr_model.fit(X_train_scaled, y_train)

LinearRegression()

In [ ]:
y_pred = lr_model.predict(X_test_scaled)

In [ ]:
r2 = r2_score(y_test, y_pred)
print("r-squared:", r2)

n = X_test.shape[0]
p = X_test.shape[1]

adjusted_r2 = 1 - ((1-r2) * (n-1) / (n-p-1))
print("adjusted r^2:", adjusted_r2)

r-squared: 0.2463684717252358
adjusted r^2: 0.19969548090311928


# Assignment 2



In [69]:
df = pd.read_csv('/content/drive/MyDrive/AI Engineer Course by Apna College/datasets/employee_turnover.csv')
df.head()

,Job_Satisfaction,Performance_Rating,Years_At_Company,Work_Life_Balance,Distance_From_Home,Monthly_Income,Education_Level,Age,Num_Companies_Worked,Employee_Role,Annual_Bonus,Training_Hours,Department,Annual_Bonus_Squared,Annual_Bonus_Training_Hours_Interaction,Employee_Turnover
0,0.562326,0.141129,0.123989,0.347583,0.330353,0.328853,0.600933,0.315990,0.768736,0.090671,0.324786,0.669193,0.602932,0.105486,0.217344,0
1,0.017041,0.559047,0.511203,0.793908,0.423550,0.553450,0.742009,0.897146,0.380035,0.601633,0.694611,0.043271,0.800761,0.482484,0.030056,0
2,0.774699,0.604371,0.798174,0.260500,0.804034,0.131800,0.775178,0.830947,0.218726,0.972936,0.153476,0.701336,0.705275,0.023555,0.107638,1
3,0.628174,0.385249,0.230104,0.516809,0.272248,0.589249,0.482409,0.090507,0.402746,0.132842,0.305973,0.549688,0.600531,0.093620,0.168190,0
4,0.799183,0.199967,0.839029,0.247927,0.341934,0.076818,0.055356,0.680860,0.923341,0.493017,0.844094,0.793751,0.664679,0.712494,0.670000,0


In [70]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1350 entries, 0 to 1349
Data columns (total 16 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   Job_Satisfaction                         1350 non-null   float64
 1   Performance_Rating                       1350 non-null   float64
 2   Years_At_Company                         1350 non-null   float64
 3   Work_Life_Balance                        1350 non-null   float64
 4   Distance_From_Home                       1350 non-null   float64
 5   Monthly_Income                           1350 non-null   float64
 6   Education_Level                          1350 non-null   float64
 7   Age                                      1350 non-null   float64
 8   Num_Companies_Worked                     1350 non-null   float64
 9   Employee_Role                            1350 non-null   float64
 10  Annual_Bonus                             1350 no

In [72]:
X = df.drop('Employee_Turnover', axis=1)
y = df['Employee_Turnover']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=43)


In [73]:
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [90]:
lr_model = LinearRegression()

lr_model.fit(X_train_scaled, y_train)

LinearRegression()

In [91]:
y_pred = lr_model.predict(X_test_scaled)

In [92]:
mse = mean_squared_error(y_test, y_pred)
print("mse:", mse)


mse: 0.10366336322034271


In [81]:
a = [0.001, 0.01, 0.1, 1, 2, 3, 5, 6, 12, 23, 45, 90, 100]

lasso_cv_model = LassoCV(alphas=a, cv=5, max_iter=1000, random_state=34)

lasso_cv_model.fit(X_train_scaled, y_train)

y_pred_lasso_cv = lasso_cv_model.predict(X_test_scaled)


In [82]:
print("The best alphe value for lassoCV: ", lasso_cv_model.alpha_)


mse = mean_squared_error(y_test, y_pred_lasso_cv)
print("mse:", mse)

r2 = r2_score(y_test, y_pred_lasso_cv)
print("r2 score:", r2)

The best alphe value:  0.01
mse: 0.1050117290080421


In [88]:
a = [0.001, 0.01, 0.1, 1, 2, 3, 5, 6, 12, 14, 13, 20, 21, 22, 19, 23, 45, 90, 100]

regit_cv_model = RidgeCV(alphas=a, cv=5)

regit_cv_model.fit(X_train_scaled, y_train)

y_pred_regdi_cv = regit_cv_model.predict(X_test_scaled)

In [89]:
print("The best alphe value for RegdiCV: ", regit_cv_model.alpha_)


mse = mean_squared_error(y_test, y_pred_regdi_cv)
print("mse:", mse)

r2 = r2_score(y_test, y_pred_regdi_cv)
print("r2 score:", r2)

The best alphe value for RegdiCV:  23.0
mse: 0.10408177270394638
r2 score: 0.5803572130901117


The best model is linear regression V as the mse is 0.103 which is less then lassoCV 105 and RegdiC 0.104.